cpp/include/cugraph/algorithms.hpp

```diff
@@ -1538,4 +1538,36 @@ void core_number(raft::handle_t const& handle,
                  size_t k_last           = std::numeric_limits<size_t>::max(),
                  bool do_expensive_check = false);
 
+/**
+ * @brief Multi-GPU Uniform Neighborhood Sampling.
+ * @tparam graph_view_t Type of graph view.
+ * @tparam index_t Type used for indexing; typically edge_t
+ * @param handle RAFT handle object to encapsulate resources (e.g. CUDA stream, communicator, and
+ * handles to various CUDA libraries) to run graph algorithms.
+ * @param graph_view Graph View object to generate NBR Sampling on.
+ * @param ptr_d_start Device array of pairs: (starting_vertex_index, rank) for the NBR Sampling.
+ * @param num_starting_vs size of starting vertex set
+ * @param h_fan_out vector of branching out (fan-out) degree per source vertex for each level
+ * parameter used for obtaining local out-degree information
+ * @param flag_replacement boolean flag specifying if random sampling is done without replacement
+ * (true); or, with replacement (false); default = true;
+ * @return tuple of tuple of device vectors and counts:
+ * ((vertex_t source_vertex, vertex_t destination_vertex, int rank, edge_t index), rx_counts)
+ */
+template <typename graph_view_t,
+          typename gpu_t,
+          typename index_t = typename graph_view_t::edge_type>
+std::tuple<std::tuple<rmm::device_uvector<typename graph_view_t::vertex_type>,
+                      rmm::device_uvector<typename graph_view_t::vertex_type>,
+                      rmm::device_uvector<gpu_t>,
+                      rmm::device_uvector<index_t>>,
+           std::vector<size_t>>
+uniform_nbr_sample(raft::handle_t const& handle,
+                   graph_view_t const& graph_view,
+                   typename graph_view_t::vertex_type const* ptr_d_start,
+                   gpu_t const* ptr_d_ranks,
```

Comment:
This is targeting a very _specific_ implementation _with_ MPI (NCCL). For this very reason alone we _should_ actually suggest the affiliation with _ranks_. In addition this is what the user (GNN consumer) wants to call them.

Reply:
FWIW, I also found the word `rank` confusing in this context at first since it wasn't clear if it was intended to be a GPU rank for use by the sampling algo itself, or a rank used by the caller (ie. a client may have their own notion of rank independent of the MG rank used by this algo). I think a more general-purpose term like `label` or `id` not only works in this case, but can also be used as a more generic mechanism for associating results to a caller's seeds. Lastly, the GNN client will likely not be directly calling this API anyway (they'll be using python), so we can make something more application-specific for them in python if necessary.